toolkit/scripts/precache.mk

```diff
@@ -0,0 +1,50 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+# Contains:
+#	- Tools to fully hydrate the rpm cache from upstream repos with the any versions that match local .spec files
+
+.PHONY: pre-cache clean-precache
+
+precache_state_dir = $(CACHED_RPMS_DIR)/precache
+precache_downloaded_files = $(precache_state_dir)/downloaded_files.txt
+precache_chroot_dir = $(precache_state_dir)/chroot
+
+$(call create_folder,$(precache_state_dir))
+
+clean-cache: clean-precache
+clean: clean-precache
+clean-precache:
+	@echo Verifying no mountpoints present in $(precache_chroot_dir)
+	$(SCRIPTS_DIR)/safeunmount.sh "$(precache_chroot_dir)" && \
+	rm -rf $(precache_chroot_dir)
+
+# We always want to run the precache script, it will decide if it needs to download anything and update the flag file if
+# it does, so add the phony target as a dependency to the flag file
+.PHONY: precache_always_run_phony
+pre-cache: $(STATUS_FLAGS_DIR)/precache.flag
+$(STATUS_FLAGS_DIR)/precache.flag: $(go-precacher) $(chroot_worker) $(rpms_snapshot) precache_always_run_phony 
+	@if [ "$(DISABLE_UPSTREAM_REPOS)" = "y" ]; then \
+		echo "ERROR: Upstream repos are disabled (DISABLE_UPSTREAM_REPOS=y), cannot precache rpms"; \
+		exit 1; \
+	fi
+	[ -f $@ ] || touch $@ # Create the flag file if it doesn't exist in case we don't download anything
+	$(go-precacher) \
+		--snapshot "$(rpms_snapshot)" \
+		--output-dir "$(rpm_cache_cache_dir)" \
+		--output-summary-file "$(precache_downloaded_files)" \
+		$(foreach url,$(PACKAGE_URL_LIST), --repo-url "$(url)") \
+		--worker-tar $(chroot_worker) \
+		--worker-dir $(precache_chroot_dir) \
+		--log-file=$(SRPM_BUILD_LOGS_DIR)/precacher.log \
+		--log-level=$(LOG_LEVEL) \
+		--cpu-prof-file=$(PROFILE_DIR)/precacher.cpu.pprof \
+		--mem-prof-file=$(PROFILE_DIR)/precacher.mem.pprof \
+		--trace-file=$(PROFILE_DIR)/precacher.trace \
+		$(if $(filter y,$(ENABLE_CPU_PROFILE)),--enable-cpu-prof) \
+		$(if $(filter y,$(ENABLE_MEM_PROFILE)),--enable-mem-prof) \
+		$(if $(filter y,$(ENABLE_TRACE)),--enable-trace) \
+		--timestamp-file=$(TIMESTAMP_DIR)/precacher.jsonl && \
+	if [ -s "$(precache_downloaded_files)" ]; then \
+		touch $@; \
+	fi
```

Comment:
If we always want to run `pre-cache`, could we just drop the status flag? Other targets depending on the pre-cache being present can depend on the `pre-cache` target as well, I believe. Or am I missing something here?
```suggestion
# We always want to run the precache script, it will decide if it needs to download anything.
.PHONY: pre-cache
pre-cache: $(go-precacher) $(chroot_worker) $(rpms_snapshot)
```

```suggestion
# We always want to run the precache script, it will decide if it needs to download anything.
.PHONY: pre-cache
pre-cache: $(go-precacher) $(chroot_worker) $(rpms_snapshot)
	@if [ "$(DISABLE_UPSTREAM_REPOS)" = "y" ]; then \
		echo "ERROR: Upstream repos are disabled (DISABLE_UPSTREAM_REPOS=y), cannot precache rpms"; \
		exit 1; \
	fi
	$(go-precacher) \
		--snapshot "$(rpms_snapshot)" \
		--output-dir "$(rpm_cache_cache_dir)" \
		--output-summary-file "$(precache_downloaded_files)" \
		$(foreach url,$(PACKAGE_URL_LIST), --repo-url "$(url)") \
		--worker-tar $(chroot_worker) \
		--worker-dir $(precache_chroot_dir) \
		--log-file=$(SRPM_BUILD_LOGS_DIR)/precacher.log \
		--log-level=$(LOG_LEVEL) \
		--cpu-prof-file=$(PROFILE_DIR)/precacher.cpu.pprof \
		--mem-prof-file=$(PROFILE_DIR)/precacher.mem.pprof \
		--trace-file=$(PROFILE_DIR)/precacher.trace \
		$(if $(filter y,$(ENABLE_CPU_PROFILE)),--enable-cpu-prof) \
		$(if $(filter y,$(ENABLE_MEM_PROFILE)),--enable-mem-prof) \
		$(if $(filter y,$(ENABLE_TRACE)),--enable-trace) \
		--timestamp-file=$(TIMESTAMP_DIR)/precacher.jsonl && \
```

Reply:
Its a very niche behavior in Make, but important. A `.PHONY` target will always be 'out of date', ie anything that depends on it will rebuild.

In this case, we always want the tool to run and make sure everything is looking good, but if we run it and decide that nothing interesting happened (ie we didn't download anything) then we shouldn't trigger any follow on steps because of this tool. That is why we check the contents of the summary file, only if the summary file has something in it do we `touch` the flag file.

If Make runs a recipe to make a file, and the file doesn't get changed, it assumes that nothing that depends on it needs to regenerate (assuming all other dependencies are also unchanged).